# Find the Bug
```go
import "log"

// customError is just an empty struct
type customError struct{}

// Error implements the error interface
func (c *customError) Error() string {
    return "Find the bug."
}

// fail returns nil values for both return types
func fail() ([]byte, *customError) {
    return nil, nil
}

func main() {
    var err error
    if _, err = fail(); err != nil {
        log.Fatal("Why did this fail?")
    }
    log.Println("No Error")
}
```
![image](https://user-images.githubusercontent.com/11031915/66178843-bd4ae200-e634-11e9-8ac3-e97edce4ead9.png)
## BUG: error interface should be used, not customError
* FIX: 
```go
func fail() ([]byte, error) {
    return nil, nil
}
``` 
## Why use pointer semantics for custom errors?
* When using value semantics, only the value of the error msg (a string) gets looked at  / compared, and string comparison is dangerous
* Interface values are valueless, and when we do comparisons btw the interface values, we compare the data stored inside of it
    - if value semantics, we compare the value itself
    - if pointer semantics, we compare the pointer